### ISSUE TITLE
Join over two columns with same name

### ISSUE THREAD

**Issue by dachafra** [ontop/ontop#325]
### Description
R2RML is not translated well to OBDA mappings when a `rr:joinCondition` is specified between two columns with the same name

### Steps to Reproduce

1. Create an R2ML mapping with a `rr:joinCondition` where the `rr:child` and `rr:parent` columns are named equal
2. Translate the R2RML mapping to OBDA mapping
3. Run a query that exploits the `rr:joinCondition`

**Expected behaviour:**  Ontop should return the result-set of the input query

**Actual behaviour:** The engine does not know which column has to take from the `rr:joinCondition` as the SQL query returns two columns that are named equal. Part of the log:
```
source query: SELECT * FROM (SELECT * FROM example1) AS child, (SELECT * FROM example2) AS parent WHERE child.id=parent.id
org.semanticweb.owlapi.reasoner.IllegalConfigurationException: it.unibz.inf.ontop.exception.InvalidMappingSourceQueriesException: Error: The source query does not provide the attribute id (variable id) required by the target atom.
Problem location: source query of the mapping assertion
```
**Reproduces how often:** Every Time

### Attached material
[example-ontop.zip](https://github.com/ontop/ontop/files/4185609/example-error-ontop.zip)

### Versions
Tested in Ontop v3.0.1 (Cli) over MySQL but also PostgreSQL

### Aditional Information
The only possible solution at this moment is to run first a translation from R2RML to OBDA mapping and edit the mapping manually which enforces the user to understand the syntax of the OBDA mappings instead of the R2RML W3C recommendation, this generates difficulties in order to use the engine

**Comment by bcogrel** [comment]
Hi David,

Thanks a lot for reporting this issue and sorry for the late reply.

Roman (@kontchakov) has done some important refactoring around this topic during the past few weeks, which should normally address this issue. We still need to test your attachment and we expect to do it soon. Note that these changes only concern the upcoming 4.0.0-beta-2 version, not 3.x, which we are not planning to fix. The code is available in the `version4` branch.

Best,
Benjamin

**Comment by bcogrel** [comment]
I adapted your test to H2 and added it to the main branch (https://github.com/ontop/ontop/commit/e66ca26aec0d9b9a7cfd24dc9c6ae60818d238ed). Thanks again!

**Comment by dachafra** [comment]
Hi @bcogrel 
I tried to reproduce my example over MySQL and I obtained the following error
```
CQC-REMOVED: EXTENSIONAL "example2"(0:id2) FROM [EXTENSIONAL "example1"(0:primarykey1,1:id2), EXTENSIONAL "example2"(0:id2)]
it.unibz.inf.ontop.owlapi.exception.OntopOWLException: it.unibz.inf.ontop.exception.OntopReformulationException: java.lang.UnsupportedOperationException: getNativeDBString is not supported by the mockup implementation
	at it.unibz.inf.ontop.owlapi.connection.impl.DefaultOntopOWLStatement.executeSelectQuery(DefaultOntopOWLStatement.java:90)
	at it.unibz.inf.ontop.cli.OntopQuery.run(OntopQuery.java:115)
	at it.unibz.inf.ontop.cli.Ontop.main(Ontop.java:18)
Caused by: it.unibz.inf.ontop.exception.OntopReformulationException: java.lang.UnsupportedOperationException: getNativeDBString is not supported by the mockup implementation
	at it.unibz.inf.ontop.answering.reformulation.impl.QuestQueryProcessor.reformulateIntoNativeQuery(QuestQueryProcessor.java:139)
	at it.unibz.inf.ontop.answering.connection.impl.QuestStatement.executeInThread(QuestStatement.java:299)
	at it.unibz.inf.ontop.answering.connection.impl.QuestStatement.execute(QuestStatement.java:182)
	at it.unibz.inf.ontop.owlapi.connection.impl.DefaultOntopOWLStatement.executeSelectQuery(DefaultOntopOWLStatement.java:85)
	... 2 more
Caused by: java.lang.UnsupportedOperationException: getNativeDBString is not supported by the mockup implementation
	at it.unibz.inf.ontop.model.term.functionsymbol.db.impl.MockupDBIsNullOrNotFunctionSymbolImpl.getNativeDBString(MockupDBIsNullOrNotFunctionSymbolImpl.java:20)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultSQLTermSerializer.lambda$serialize$5(DefaultSelectFromWhereSerializer.java:361)
	at java.util.Optional.map(Optional.java:215)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultSQLTermSerializer.serialize(DefaultSelectFromWhereSerializer.java:361)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.lambda$visit$0(DefaultSelectFromWhereSerializer.java:76)
	at java.util.Optional.map(Optional.java:215)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:76)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer$DefaultRelationVisitingSerializer.visit(DefaultSelectFromWhereSerializer.java:45)
	at it.unibz.inf.ontop.generation.algebra.impl.SelectFromWhereWithModifiersImpl.acceptVisitor(SelectFromWhereWithModifiersImpl.java:99)
	at it.unibz.inf.ontop.generation.serializer.impl.DefaultSelectFromWhereSerializer.serialize(DefaultSelectFromWhereSerializer.java:38)
	at it.unibz.inf.ontop.generation.DefaultSQLIQTree2NativeNodeGenerator.generate(DefaultSQLIQTree2NativeNodeGenerator.java:52)
	at it.unibz.inf.ontop.answering.reformulation.generation.impl.SQLGeneratorImpl.generateNativeNode(SQLGeneratorImpl.java:140)
	at it.unibz.inf.ontop.answering.reformulation.generation.impl.SQLGeneratorImpl.generateSourceQuery(SQLGeneratorImpl.java:84)
	at it.unibz.inf.ontop.answering.reformulation.impl.QuestQueryProcessor.generateExecutableQuery(QuestQueryProcessor.java:148)
	at it.unibz.inf.ontop.answering.reformulation.impl.QuestQueryProcessor.reformulateIntoNativeQuery(QuestQueryProcessor.java:121)
	... 5 more
`
``

**Comment by bcogrel** [comment]
Hi @dachafra 

Have you tried Ontop 4.0.2? It contains a bugfix that could solve your issue I guess : https://github.com/ontop/ontop/commit/b9046e63a07d2ddec24e8ff33bbec46bea1ea6b9

Prior to 4.0.2, Ontop was not always that it was dealing with a MySQL database when the value for `jdbc.driver`  was not   `com.mysql.jdbc.Driver`. In such a case, it was failing back to the wrong default factories, the mockup ones  (instead of the default SQL ones), hence your exception.

Let us know whenever 4.0.2 address your issue or not.

Best,
Benjamin

**Comment by dachafra** [comment]
Solved! thanks @bcogrel, I was trying with 4.0.0 ;-)
